### PR TITLE
[cmake] allow set extension sources in cmake

### PR DESCRIPTION
**Pull request by gjc13**

### CODE REVIEW

File: src/core/CMakeLists.txt [hunk 1]
```diff
@@ -215,8 +215,8 @@ target_include_directories(openthread-ftd PUBLIC ${OT_PUBLIC_INCLUDES} PRIVATE $
 target_include_directories(openthread-mtd PUBLIC ${OT_PUBLIC_INCLUDES} PRIVATE ${COMMON_INCLUDES})
 target_include_directories(openthread-radio PUBLIC ${OT_PUBLIC_INCLUDES} PRIVATE ${COMMON_INCLUDES})
 
-target_sources(openthread-ftd PRIVATE ${COMMON_SOURCES})
```
Comment (bukepo) [inline src/core/CMakeLists.txt:218]: define OT_EXTENSION_SOURCES by command `set()` first?

File: src/core/CMakeLists.txt [hunk 2]
```diff
@@ -215,6 +215,12 @@ target_include_directories(openthread-ftd PUBLIC ${OT_PUBLIC_INCLUDES} PRIVATE $
 target_include_directories(openthread-mtd PUBLIC ${OT_PUBLIC_INCLUDES} PRIVATE ${COMMON_INCLUDES})
 target_include_directories(openthread-radio PUBLIC ${OT_PUBLIC_INCLUDES} PRIVATE ${COMMON_INCLUDES})
 
+
+set(OT_EXTENSION_SOURCE "" CACHE STRING "set extension source for OpenThread")
+if(OT_EXTENSION_SOURCE)
+  target_compile_definitions(ot-config INTERFACE "OPENTHREAD_ENABLE_VENDOR_EXTENSION=1")
+  list(APPEND COMMON_SOURCES ${PROJECT_SOURCE_DIR}/${OT_EXTENSION_SOURCE})
+endif()
```
Comment (bukepo) [inline src/core/CMakeLists.txt:223]: The source path can be an absolute path.
```suggestion
set(OT_VENDOR_EXTENSION "" CACHE STRING "specify a C++ source file built as part of OpenThread core library")
if(OT_VENDOR_EXTENSION)
  target_compile_definitions(ot-config INTERFACE "OPENTHREAD_ENABLE_VENDOR_EXTENSION=1")
  list(APPEND COMMON_SOURCES ${OT_VENDOR_EXTENSION})
endif()
```

**Comment by bukepo** [comment]
Could you add a test which set the vendor extension to the example(src/core/common/extension_example.cpp) like script/check-simulation-build-autotools?

**Comment by gjc13** [comment]
> Could you add a test which set the vendor extension to the example(src/core/common/extension_example.cpp) like script/check-simulation-build-autotools?

Done, see `script/check-simulation-build-cmake`

**Review by bukepo — APPROVED** [review]
LGTM:+1:

**Pull request merged**